src/client/Microsoft.Identity.Client/Platforms/Android/AndroidBroker.cs

```diff
@@ -0,0 +1,174 @@
+        private Activity _activity;
+
+
+        public AndroidBroker(ICoreLogger logger)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _brokerHelper = new AndroidBrokerHelper(Application.Context, logger);
+        }
+
+        public bool CanInvokeBroker(CoreUIParent uiParent)
+        {
+            bool canInvoke = _brokerHelper.CanSwitchToBroker();
+            _logger.Verbose("Can invoke broker? " + canInvoke);
+
+            _activity = uiParent?.Activity;
+            return canInvoke;
+        }
+
+        public async Task<MsalTokenResponse> AcquireTokenUsingBrokerAsync(Dictionary<string, string> brokerPayload)
+        {
+            _androidBrokerTokenResponse = null;
+            _readyForResponse = new SemaphoreSlim(0);
+            _correlationId = AndroidBrokerHelper.GetValueFromBrokerPayload(brokerPayload, BrokerParameter.CorrelationId);
+            //Need to disable warning for non awaited async call.
+            try
+            {
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+                new TaskFactory().StartNew(() => AcquireTokenInternalAsync(brokerPayload));
+#pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+            }
+            catch (Exception ex)
+            {
+                _logger.Error("Broker Operation Failed to complete.");
+                if (ex is MsalException)
+                    throw;
+                else
+                    throw new MsalException(MsalError.AndroidBrokerOperationFailed, ex.Message, ex);
+            }
+
+            await _readyForResponse.WaitAsync().ConfigureAwait(false);
+            return _androidBrokerTokenResponse;
+        }
+
+        private async Task AcquireTokenInternalAsync(IDictionary<string, string> brokerPayload)
+        {
+            _brokerHelper.InitiateBrokerHandshake(_activity);
+
+            Context mContext = Application.Context;
+
+            brokerPayload[BrokerParameter.BrokerAccountName] = AndroidBrokerHelper.GetValueFromBrokerPayload(brokerPayload, BrokerParameter.LoginHint);
+
+            // Don't send silent background request if account information is not provided
+            if (!string.IsNullOrEmpty(AndroidBrokerHelper.GetValueFromBrokerPayload(brokerPayload, BrokerParameter.BrokerAccountName)) ||
+                !string.IsNullOrEmpty(AndroidBrokerHelper.GetValueFromBrokerPayload(brokerPayload, BrokerParameter.Username)))
+            {
+                _logger.Verbose("User is specified for silent token request. Starting silent broker request");
+                string silentResult = _brokerHelper.GetBrokerAuthTokenSilently(brokerPayload, _activity);
+                _androidBrokerTokenResponse = CreateMsalTokenResponseFromResult(silentResult);
+                _readyForResponse.Release();
```

Comment:
This looks like a race condition. Any code that runs after line will potentially be ignored because when you relesae the semaphore the other thread will continue execution (after line 78) and will return null (line 80) or even worse - return an old result.